### DISCUSSION TITLE
Combine outputs in test epochs when using DDP

### DISCUSSION THREAD

**Discussion by WouterDurnez** [Lightning-AI/pytorch-lightning#11086] [DDP / multi-GPU / multi-node] [answered]
I'm training a model across two GPUs on patient data (id). In my test steps, I output dictionaries, which contain the id, as well as all the metrics. I store these (a list with a dict per id) at the end of the test epoch, so I can later on statistically evaluate model performances.

I'm experiencing a problem with the test step, however. 

    # Test step
    def test_step(self, batch, batch_idx):

        # Get new input and predict, then calculate loss
        x, y, id = batch["input"], batch["target"], batch["id"]

        # Infer and time inference
        start = time()
        y_hat = self.test_inference(x, self, **self.test_inference_params)
        end = time()

        # Calculate metrics
        id = id[0] if len(id) == 1 else tuple(id)

        # Output dict with duration of inference
        output = {"id": id, "time": end - start}

        # Add other metrics to output dict
        for m, pars in zip(self.metrics, self.metrics_params):

            metric_value = m(y_hat, y, **pars)

            if hasattr(metric_value, "item"):
                metric_value = metric_value.item()

            output[f"test_{m.__name__}"] = metric_value

        return output

    # Test epoch end (= test end)
    def test_epoch_end(self, outputs):

        # Go over outputs and gather
        self.test_results = outputs     #self.all_gather(outputs)

I hadn't considered this before (as I'm used to training on a single GPU), but the test_results attribute now only contains half of the outputs (one half per process). So when my main script reaches this section, only half the output is effectively stored:

    log("Evaluating model.")
    trainer.test(model=model,
                 dataloaders=brats.val_dataloader())
    results = model.test_results

    # Save test results
    log("Saving results.")
    np.save(file=join(result_dir, f'{model_name}_v{version}_fold{fold_index}.npy'), arr=results)

I have read about the `self.all_gather` method, but I'm not sure it suits my needs. I want to merge the lists, not reduce anything. Also, they're not Tensors, but dicts. How can I store all dicts across both DDP processes?

**Comment by rohitgr7** [comment, marked as answer]
all_gather is different from all_reduce. It doesn't do any math operation here.
sort of like:
```
all_gather -> collect outputs from all devices
all_reduce -> in general, collect outputs from all devices and reduce (apply a math op)
```
all_gather isn't working for you?

**Reply by WouterDurnez** [reply]
I currently got this:

`RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.`

...which I can perhaps solve by adding `.detach()` to the `self.all_gather(outputs)` line? But it's not what I wanted/expected (just an extended list of dicts).

**Reply by rohitgr7** [reply]
I might be wrong here but some of the contents inside your output are scalars and not tensors. For eg: `time`, metric value. I think all_gather can only collect tensors.

**Reply by WouterDurnez** [reply]
Yes, you're exactly right. All of them are scalars, in fact (I take out the value if a Tensor is returned). So are you saying it's impossible to merge a list of dicts, in which the values are scalars? And actually, the 'id' is a string :)

**Reply by rohitgr7** [reply]
I have mostly worked with tensors tbh, so not sure about strings. 
maybe @awaelchli has some idea regarding this.

**Reply by awaelchli** [reply]
Can you convert the contents to tensors (using `torch.tensor(x, device=self.device`)) and try again with `self.all_gather`?

**Reply by WouterDurnez** [reply]
I'll give that a go, thanks. I'll update when I have results.

**Reply by WouterDurnez** [reply]
Seems like it all worked! Thanks @awaelchli and @rohitgr7!

**Reply by icoz69** [reply]
hi, it seems that test_epoch_end is also runned on each GPU node, right? Does it mean each node gathers the outputs from all nodes and does the same thing?

**Reply by rohitgr7** [reply]
yes @icoz69,
not just on each GPU node, but on each device. To add some restrictions, you can use `trainer.is_global_zero` so that it will collect only on one of the device across nodes.
```py
def test_epoch_end(...):
    if self.trainer.is_global_zero:
        # collect or gather
```

**Comment by Jerzy97** [comment]
Hi @WouterDurnez,
I am in a very similar situation and I have a follow up. How did you solve the problem of keeping track of the patient ID?
Did you map every ID (string) to a unique integer?

Thanks for your hints

**Reply by WouterDurnez** [reply]
Hi @Jerzy97. Yes, in my case, the ids were already as simple as `f"BraTS{patient integer id}"`, so I just parsed the integer. But you can map just as easily, sure. That solved it for me.

**Reply by Jerzy97** [reply]
I see - thanks for your response!

**Comment by mgwillia** [comment]
I would think all_gather_object would be optimal given the data types, no?
